fdroidserver 0.4.0 (new formula)

eighthave opened the pull request:
Tools for creating and managing Android app repositories for F-Droid
https://f-droid.org

mikemcquaid commented on Library/Formula/fdroidserver.rb:
```diff
@@ -0,0 +1,99 @@
+    For more info:
+       https://sites.google.com/a/android.com/tools/tech-docs/configuration/osx-jdk
+    EOS
+  end
```
Just do `depends_on :java` and `depends_on "android_sdk"`

eighthave replied:
I left out `depends_on :java` because it wasn't clear to me that that would also install the JDK.  Plus that might install the JDK6, which is tool old for `fdroidserver`.  Oracle seems to have dropped its JDK7 downloads and only has JDK8 now.

mikemcquaid replied:
`depends_on :java => "1.7"`

eighthave replied:
Makes sense.  I couldn't find any docs on the `depends_on :java` stuff, but I think this sounds like it addresses the issues.

mikemcquaid commented on Library/Formula/fdroidserver.rb:
```diff
@@ -0,0 +1,99 @@
+    else
+      puts "Tests require the env var ANDROID_HOME to point to the Android SDK!"
+      puts "Skipping the test."
+    end
```
This test should instead be modified to find the Android SDK based on `depends_on "android_sdk"` and test that. We can't rely on this variable being set.

eighthave replied:
The problem is that this actually depends on the `build-tools` section of the Android SDK, which is not automatically installed when doing `brew install android-sdk`.  Also, I think a lot of people get their Android SDK via the Android Studio bundle, even with brew installed, that's why I want to make it "recommended" rather than required.

mikemcquaid replied:
`depends_on "android-sdk" => :recommended` is what we want.

mikemcquaid commented on Library/Formula/fdroidserver.rb:
```diff
@@ -0,0 +1,99 @@
+    if ENV["ANDROID_HOME"]
+      fdroidrepo = testpath/"fdroid"
+      fdroidrepo.mkdir()
+      fdroidrepo.cd do
```
Not clear why this needs to be done in a created subdirectory

eighthave replied:
the fdroid tools prefer working in a directory called `fdroid`, which in turn matches how the published repo looks on the webserver.

mikemcquaid replied:
Ok. In that case you can use `mkdir testpath/"fdroid" do`

dunn commented:
We're getting a build error on all three platforms:
```
_imagingft.c:31:10: fatal error: 'ft2build.h' file not found
#include <ft2build.h>
         ^
```
And there are two audit warnings:
```
==> audit problems
fdroidserver:
 * Description is too long. "name: desc" should be less than 80 characters.
Length is calculated as fdroidserver + desc. (currently 82)
 * Dependency 'libjpeg' is an alias; use the canonical name 'jpeg'.
```

eighthave commented:
I fixed the description and changed 'libjpeg' to 'jpeg'.  About the ft2build.h error, that one is stumping me.  Python Pillow needs freetype 2 to build, and that is what the `freetype` brew package seems to provide.  So it seems like there is a bug somewhere because the Pillow build cannot find that freetype header. I bumped the Pillow version to the latest (v2.9.0) in the hopes that it might help.  The odd thing is that it was working on my machine when I first submitted this formula here.

DomT4 commented:
You may need to do something like 

```ruby
ENV.append_to_cflags "-I#{Formula["freetype"].opt_include}/freetype2"
```

eighthave commented:
@DomT4 that did it, thanks!  Now the last remaining issue is the tests.  In order to actually run the tests, I'd need to run the Android SDK installer to install the "SDK Build-tools" package, i.e.
```
android update sdk --no-ui --filter tool
```

Is that something that's OK to do in the tests?

eighthave commented:
I added this to the test section, and it worked!

```ruby
  test do
    exec "echo yes | android -v update sdk --no-ui --force --filter platform-tools,build-tools-22.0.1"
    ...
```

DomT4 commented:
> android update sdk --no-ui --filter tool

It depends on how large a download it is, but as long as it isn't huge and doesn't take too long, sure.

You could do something like:

```ruby
test do
  system bin/"android", "update", "sdk", "--no-ui", "--filter tool"
  assert File.exist?("expected_downloaded_file.file")
end
```

mikemcquaid commented:
Yeh, I'm up for including it :+1:

mikemcquaid commented:
You could even add that to the `android-sdk` formula and install it by default for everyone?

eighthave commented:
I just found a missing dependency for apache-libcloud, `backports.ssl_match_hostname`, and added that to this recipe.  This is ready to be merged! :-)

mikemcquaid commented on Library/Formula/fdroidserver.rb:
```diff
@@ -0,0 +1,104 @@
+
+  test do
+    # fdroid requires "SDK Platform-tools" and "SDK Build-tools"
+    exec "echo yes | android -v update sdk --no-ui --force --filter platform-tools,build-tools-22.0.1"
```
Can you try to use `pipe_output` instead of using a pipe here? Otherwise `system` rather than `exec` here.

eighthave replied:
I can try, but I really don't know ruby at all, can't you tell? ;-)

bfontaine replied:
@eighthave Instead of using `exec "echo something | some command here"` use `pipe_output("some command here", "something", 0)`.

mikemcquaid replied:
:+1:

eighthave replied:
Ok, I updated it to use `pipe_output()`

eighthave replied:
Well, the `pipe_output()` fails where the `exec()` worked just fine.  This worked:

bfontaine replied:
@eighthave when you do it by hand, what’s the exit code?

```sh
echo y | android -v update sdk --no-ui --force --filter platform-tools,build-tools-22.0.1
echo $?
```

If it’s not `0`, replace the last argument of `pipe_output` with the correct number. [See here](https://github.com/Homebrew/homebrew/blob/4fbd40a4ebace7bf0809effd62a4fa3a1993d899/Library/Formula/moreutils.rb#L64) for an example.

mikemcquaid commented on Library/Formula/fdroidserver.rb:
```diff
@@ -0,0 +1,104 @@
+    base path of the Android SDK is set in the standard environment variable
+    ANDROID_HOME.  To install them from the command line, run:
+    android update sdk --no-ui --filter platform-tools,build-tools-22.0.1
+    EOS
```
It'd be good to fix the android-sdk formula to maybe just install the platform tools and also if you can make this formula autoinstall (as you do in `test do`) in `postinstall`

eighthave replied:
Build-tools and Platform-tools are both required to actually run builds.  Why install the android-sdk install in `postinstall` and not as a regular dependency?

As for modifying the android-sdk package, I do not feel like I'm a good candidate since I don't do Android development on OSX and don't use that package beyond this dependency.  I don't have any info on the use cases for how people use it, so I feel like I could easily be messing things up for people who use it.

eighthave replied:
Also, it feels somehow wrong to me for this fdroidserver formula to modify something outside of its domain, as in installing things in the Android SDK.  But maybe that's thinking too Debian-ish and brew users expect that?

mikemcquaid replied:
Sorry, I'm saying the `android-sdk` formula could/should be modified to install these things. I really don't think it'll mess anything up.

eighthave replied:
I looked into doing this.  There are a lot of notes and caveats so its a complicated thing.  I don't currently have time to take on that project.  For example:

    # The 'android' tool insists on deleting #{prefix}/platform-tools
    # and then installing the new one. So it is impossible for us to redirect
    # the SDK location to var so that the platform-tools don't have to be
    # freshly installed EVERY DANG time the base SDK updates.

bfontaine commented on Library/Formula/fdroidserver.rb:
```diff
@@ -0,0 +1,105 @@
+  test do
+    # fdroid requires "SDK Platform-tools" and "SDK Build-tools"
+    output = pipe_output("android -v update sdk --no-ui --force --filter platform-tools,build-tools-22.0.1", "y\n", 0)
+    print output
```
Please remove this line.

eighthave commented:
This `pipe_output()` change seems to break things on the jenkins build.  This recipe builds and tests fine on my Mavericks laptop, and I can't seem to figure out how to get the output from the `android` command on the jenkins build.  How about I switch back to the `exec` call that worked?

bfontaine commented on Library/Formula/fdroidserver.rb:
```diff
@@ -0,0 +1,104 @@
+
+  test do
+    # fdroid requires "SDK Platform-tools" and "SDK Build-tools"
+    pipe_output("android --verbose update sdk --no-ui --all --filter platform-tools,build-tools-23.0.1", "y\n")
```
Use `"#{Formula["android-sdk"].bin}/android ...` instead of `"android ...".

bfontaine commented on Library/Formula/fdroidserver.rb:
```diff
@@ -0,0 +1,104 @@
+    pipe_output("android --verbose update sdk --no-ui --all --filter platform-tools,build-tools-23.0.1", "y\n")
+    # fdroid prefers to work in a dir called 'fdroid'
+    mkdir testpath/"fdroid" do
+      ENV["ANDROID_HOME"] = "/usr/local/opt/android-sdk"
```
Use `Formula["android-sdk"].opt_prefix`

eighthave commented:
I made the changes that @bfontaine suggested and now it builds on the travis 10.9 and 10.10, but the tests fail on 10.11.  So there is progress!

Looks like the `android` command failed to install *build-tools-23.0.1*, but there is no debug log in the travis output, even though the command is run using `--verbose`.  Seems the test setup might still be a bit brittle.

Perhaps switching back to `exec` will work? It did before...

mikemcquaid commented:
@eighthave Once #46575 is merged you should be able to just apply:

```patch
--- a/Library/Formula/fdroidserver.rb
+++ b/Library/Formula/fdroidserver.rb
@@ -88,8 +88,6 @@ class Fdroidserver < Formula
   end
 
   test do
-    # fdroid requires "SDK Platform-tools" and "SDK Build-tools"
-    pipe_output("#{Formula["android-sdk"].bin}/android --verbose update sdk --n
o-ui --all --filter platform-tools,build-tools-23.0.1", "y\n")
```
and have this work (at least it does for me locally).

bfontaine commented:
@mikemcquaid You seem to have a newline issue in your diff :wink:

mikemcquaid commented:
@bfontaine Whoops, thanks, fixed!

bfontaine commented:
ping re Mike’s comment?

eighthave commented:
Ok, made the change from mike, the travis tests failed the first time saying it couldn't find build-tools again, but then the second time they worked.  On jenkins, only the el_capitan test failed saying 

```
==> audit problems
fdroidserver:
 * The homepage is not reachable (curl exit code 22)
```

So you probably want to increase the timeout on that test.

mikemcquaid commented:
@eighthave Looks like the test is still failing due to the build tools being missing: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/41718/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/test_fdroidserver/

Not sure the best way around that? Maybe copy the `android-sdk` installation inside the `testpath`, install the build tools and then run the test as-is?

eighthave commented:
How about I remove the tests and get this merged, then come back around to the tests in an update.

mikemcquaid commented:
@eighthave Sounds good but rather than remove replace with a `fdroid --version` or similar (or if even that isn't doable: a `File.exist? bin/"fdroid"`)

eighthave commented:
Ok, I added some simpler tests and they pass on travis.  I'm still getting the errand false negative result about the home page, this time on 10.9:
```
fdroidserver:

 * The homepage is not reachable (curl exit code 28)
```
Perhaps that needs multiple retries and a long timeout.  I find that CI systems often have crappy internet response.

DomT4 commented:
Travis is ignorable on that. Seems to be having some connectivity issues this week. You passed Jenkins fine.

mikemcquaid commented:
Thanks for this and your patience @eighthave! Got there in the end!
